R/epidist.R

```diff
@@ -704,3 +704,129 @@ generate.epidist <- function(x, times, ...) {
   }
   out
 }
+
+#' Discretises a continuous distribution in an `epidist` object
+#'
+#' @details Converts the S3 distribution object in an `epidist` from continuous
+#' (using an object from the {distributional} package) to a discretised
+#' distribution (using an object from the {distcrete} package).
+#'
+#' @param x An `epidist` object
+#' @param ... further arguments passed to or from other methods
+#'
+#' @return An `epidist` object
+#' @export
+#'
+#' @examples
+#' ebola_incubation <- epidist(
+#'   disease = "ebola",
+#'   epi_dist = "incubation_period",
+#'   prob_distribution = "gamma",
+#'   prob_distribution_params = c(shape = 1, scale = 1)
+#' )
+#' discretise(ebola_incubation)
+discretise <- function(x, ...) {
+  UseMethod("discretise")
+}
+
+#' @rdname discretise
+#' @export
+discretise.epidist <- function(x, ...) {
```

Comment:
Worth flagging in function description (or somewhere else) that distributional functions may no longer apply to distcrete objects (e.g. `generate(onset_to_death_ebola$prob_dist,10)` vs
`generate(epiparameter::discretise(onset_to_death_ebola)$prob_dist,10)`? Not sure what relevant syntax is for same random draw operation with `distcrete` (and corresponding pdf/pmf functions).

Reply:
The distribution functions in {epiparameter} are setup to handle both discretised and non-discretised distributions:

``` r
library(epiparameter)
library(distributional)
ebola_si <- epidist(
  disease = "ebola", 
  epi_dist = "serial_interval", 
  prob_distribution = "gamma", 
  prob_distribution_params = c(shape = 1, scale = 1)
)
#> Citation cannot be created as either author, year or DOI is missing
ebola_si
#> Disease: ebola
#> Pathogen: NA
#> Epi Distribution: serial interval
#> Study: No citation available
#> Distribution: gamma
#> Parameters:
#>   shape: 1.000
#>   rate: 1.000

cdf(ebola_si, q = 50)
#> [1] 1
generate(ebola_si, times = 10)
#>  [1] 0.2634135 1.1430287 1.0172865 0.4948622 1.4337434 0.8281407 0.5926481
#>  [8] 0.1577783 0.1523971 0.3412915
density(ebola_si, at = 0.5)
#> [1] 0.6065307
quantile(ebola_si, p = 0.5)
#> [1] 0.6931472

ebola_si <- discretise(ebola_si)
ebola_si
#> Disease: ebola
#> Pathogen: NA
#> Epi Distribution: serial interval
#> Study: No citation available
#> Distribution: discrete gamma
#> Parameters:
#>   shape: 1.000
#>   scale: 1.000

cdf(ebola_si, q = 50)
#> [1] 1
generate(ebola_si, times = 10)
#>  [1] 2 2 2 1 1 1 2 1 2 5
density(ebola_si, at = 0.5)
#> [1] 0.3934693
quantile(ebola_si, p = 0.5)
#> [1] 0
```

<sup>Created on 2023-04-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

But I definitely agree we need to make this clear in the documentation and vignettes.